### PR TITLE
test: use fixture for gRPC metadata tests

### PR DESCRIPTION
**Pull request by coryan**
This is needed to complete the next import into Google's code
repository.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8348)

### CODE REVIEW

File: google/cloud/bigtable/async_row_reader_test.cc [hunk 1]
```diff
@@ -69,13 +69,13 @@ class TableAsyncReadRowsTest : public bigtable::testing::TableTestFixture {
...
     EXPECT_CALL(*client_, PrepareAsyncReadRows)
-        .WillOnce([&reader, request_expectations_ptr](
+        .WillOnce([this, &reader, request_expectations_ptr](
                       grpc::ClientContext* context,
                       btproto::ReadRowsRequest const& r,
                       grpc::CompletionQueue*) {
-          EXPECT_STATUS_OK(
-              IsContextMDValid(*context, "google.bigtable.v2.Bigtable.ReadRows",
-                               google::cloud::internal::ApiClientHeader()));
+          EXPECT_STATUS_OK(validate_metadata_fixture_.IsContextMDValid(
```
Comment (dbolduc) [inline google/cloud/bigtable/async_row_reader_test.cc:76]: there is only one call. Maybe we just define `ValidateMetadataFixture fixture;` in here?

File: google/cloud/bigtable/mutation_batcher_test.cc [hunk 2]
```diff
@@ -167,12 +167,12 @@ class MutationBatcherTest : public bigtable::testing::TableTestFixture {
...
 
       EXPECT_CALL(*client_, PrepareAsyncMutateRows)
-          .WillOnce([reader, exchange](grpc::ClientContext* context,
-                                       btproto::MutateRowsRequest const& r,
-                                       grpc::CompletionQueue*) {
+          .WillOnce([this, reader, exchange](
+                        grpc::ClientContext* context,
+                        btproto::MutateRowsRequest const& r,
+                        grpc::CompletionQueue*) {
             EXPECT_STATUS_OK(IsContextMDValid(
```
Comment (dbolduc) [inline google/cloud/bigtable/mutation_batcher_test.cc:174]: there is only one call. Maybe we just define `ValidateMetadataFixture fixture;` in here?

(I am saying this just for consistency. I do not care if you actually make this change.)

File: google/cloud/bigtable/table_admin_legacy_test.cc [hunk 3]
```diff
@@ -95,159 +95,190 @@ class TableAdminTest : public ::testing::Test {
...
+  CreateListTablesMock(std::string const& expected_token,
+                       std::string const& returned_token,
+                       std::vector<std::string> const& table_names) {
+    return [this, expected_token, returned_token, table_names](
+               grpc::ClientContext* context,
+               btadmin::ListTablesRequest const& request,
+               btadmin::ListTablesResponse* response) {
+      EXPECT_STATUS_OK(IsContextMDValid(
+          *context, "google.bigtable.admin.v2.BigtableTableAdmin.ListTables",
+          google::cloud::internal::ApiClientHeader()));
```
Comment (dbolduc) [inline google/cloud/bigtable/table_admin_legacy_test.cc:124]: This is setting the `resource_name` to `google::cloud::internal::ApiClientHeader()`... I'm guessing you don't want that.

File: google/cloud/bigtable/table_admin_legacy_test.cc [hunk 4]
```diff
@@ -95,159 +95,190 @@ class TableAdminTest : public ::testing::Test {
...
+  CreateListBackupsMock(std::string const& expected_token,
+                        std::string const& returned_token,
+                        std::vector<std::string> const& backup_names) {
+    return [this, expected_token, returned_token, backup_names](
+               grpc::ClientContext* context,
+               btadmin::ListBackupsRequest const& request,
+               btadmin::ListBackupsResponse* response) {
+      EXPECT_STATUS_OK(IsContextMDValid(
+          *context, "google.bigtable.admin.v2.BigtableTableAdmin.ListBackups",
+          google::cloud::internal::ApiClientHeader()));
```
Comment (dbolduc) [inline google/cloud/bigtable/table_admin_legacy_test.cc:224]: This is setting the `resource_name` to `google::cloud::internal::ApiClientHeader()`... I'm guessing you don't want that.

File: google/cloud/bigtable/table_admin_legacy_test.cc [hunk 5]
```diff
@@ -1229,22 +1248,22 @@ TEST_F(TableAdminTest, AsyncWaitForConsistencySimple) {
...
+              *context,
+              "google.bigtable.admin.v2.BigtableTableAdmin.CheckConsistency"));
+          EXPECT_EQ(
+              "projects/the-project/instances/test-instance/tables/test-table",
+              request.name());
+          // This is safe, see comments in MockAsyncResponseReader.
+          return std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<
+              btadmin::CheckConsistencyResponse>>(r.get());
+        };
+      };
```
Comment (dbolduc) [inline google/cloud/bigtable/table_admin_legacy_test.cc:1267]: GitHub diff UI won't let me comment on the line, but there is another occurrence in this file we should delete:

https://github.com/googleapis/google-cloud-cpp/blob/a93026f235e99ce00daeed0e710c589c0a110fe0/google/cloud/bigtable/table_admin_legacy_test.cc#L1343

File: google/cloud/bigtable/table_apply_test.cc [hunk 6]
```diff
@@ -26,29 +26,43 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
...
+  using MutateRowMock =
+      std::function<grpc::Status(grpc::ClientContext* context,
+                                 google::bigtable::v2::MutateRowRequest const&,
+                                 google::bigtable::v2::MutateRowResponse*)>;
+
+  MutateRowMock CreateMutateRowMock(grpc::Status const& status) {
+    return [this, status](grpc::ClientContext* context,
+                          google::bigtable::v2::MutateRowRequest const&,
+                          google::bigtable::v2::MutateRowResponse*) {
+      EXPECT_STATUS_OK(
```
Comment (dbolduc) [inline google/cloud/bigtable/table_apply_test.cc:52]: same indifferent comment about a single use `ValidateMetadataFixture fixture;`

File: google/cloud/bigtable/table_check_and_mutate_row_test.cc [hunk 7]
```diff
@@ -25,29 +25,42 @@ namespace bigtable {
...
+  using MockCheckAndMutate = std::function<grpc::Status(
+      grpc::ClientContext* context,
+      google::bigtable::v2::CheckAndMutateRowRequest const&,
+      google::bigtable::v2::CheckAndMutateRowResponse*)>;
+
+  MockCheckAndMutate CreateMockCheckAndMutate(grpc::Status const& status) {
+    return [this, status](grpc::ClientContext* context,
+                          google::bigtable::v2::CheckAndMutateRowRequest const&,
+                          google::bigtable::v2::CheckAndMutateRowResponse*) {
+      EXPECT_STATUS_OK(IsContextMDValid(
```
Comment (dbolduc) [inline google/cloud/bigtable/table_check_and_mutate_row_test.cc:50]: same. change if you feel like it.

File: google/cloud/bigtable/table_readmodifywriterow_test.cc [hunk 8]
```diff
@@ -270,7 +274,8 @@ TEST_F(TableReadModifyWriteTest, UnrecoverableFailureTest) {
       .WillRepeatedly([](grpc::ClientContext* context,
                          google::bigtable::v2::ReadModifyWriteRowRequest const&,
                          google::bigtable::v2::ReadModifyWriteRowResponse*) {
-        EXPECT_STATUS_OK(IsContextMDValid(
+        ::google::cloud::testing_util::ValidateMetadataFixture fixture;
```
Comment (dbolduc) [inline google/cloud/bigtable/table_readmodifywriterow_test.cc:277]: here, I thought we'd use the member variable `validate_metadata_fixture_`. (again, I don't care. I am going to stop commenting on these things)

File: google/cloud/spanner/internal/metadata_spanner_stub_test.cc [hunk 9]
```diff
@@ -60,15 +59,27 @@ class MetadataSpannerStubTest : public ::testing::Test {
...
   template <typename MockCall, typename Request, typename MemberFunction>
   void ExpectSession(MockCall& call, Request const&,
                      std::string const& rpc_name,
                      MemberFunction member_function) {
     call.WillOnce(
         [this, rpc_name](grpc::ClientContext& context, Request const&) {
-          EXPECT_STATUS_OK(
-              IsContextMDValid(context, "google.spanner.v1.Spanner." + rpc_name,
-                               expected_api_client_header_, db_.FullName()));
+          EXPECT_STATUS_OK(validate_metadata_fixture_.IsContextMDValid(
```
Comment (dbolduc) [inline google/cloud/spanner/internal/metadata_spanner_stub_test.cc:80]: why don't we use the helper method here?

**Review by dbolduc — CHANGES_REQUESTED** [review]
`bigtable/table_admin_legacy_test.cc` "needs" changes. Anything else is optional. I am not thrilled that the code is inconsistent but I am not going to do anything about it.

If you want to really save lines, you could define these things once in `validate_metadata.*` instead of in each test.
```cc
  Status IsContextMDValidNormal(grpc::ClientContext& context, std::string const& method) {
    return IsContextMDValid(context, method, google::cloud::internal::ApiClientHeader());
  }
  Status IsContextMDValidNormal(grpc::ClientContext& context, std::string const& method, std::string const& resource_name) {
    return IsContextMDValid(context, method, google::cloud::internal::ApiClientHeader(), resource_name);
  }

  Status IsContextMDValidGenerator(grpc::ClientContext& context, std::string const& method) {
    return IsContextMDValid(context, method, google::cloud::internal::ApiClientHeader("generator"));
  }
  Status IsContextMDValidGenerator(grpc::ClientContext& context, std::string const& method, std::string const& resource_name) {
    return IsContextMDValid(context, method, google::cloud::internal::ApiClientHeader("generator"), resource_name);
  }
```

File: google/cloud/spanner/internal/metadata_spanner_stub_test.cc (hunk 9 above)
Comment (coryan) [reply google/cloud/spanner/internal/metadata_spanner_stub_test.cc]: Fixed.

File: google/cloud/bigtable/table_readmodifywriterow_test.cc (hunk 8 above)
Comment (coryan) [reply google/cloud/bigtable/table_readmodifywriterow_test.cc:276]: You are right, fixed.

File: google/cloud/bigtable/async_row_reader_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/bigtable/async_row_reader_test.cc:76]: The function around this is called many times though, I am happy with it as-is.

File: google/cloud/bigtable/mutation_batcher_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/bigtable/mutation_batcher_test.cc:174]: Ditto, the surrounding function is called many times.

I checked again, and the places where I use a local fixture are (lucky for me) only called once.

File: google/cloud/bigtable/table_admin_legacy_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/bigtable/table_admin_legacy_test.cc]: Fixed.  I think that says something about our tests.

File: google/cloud/bigtable/table_admin_legacy_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/bigtable/table_admin_legacy_test.cc]: Fixed.

File: google/cloud/bigtable/table_admin_legacy_test.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/bigtable/table_admin_legacy_test.cc:1265]: Fixed .

File: google/cloud/bigtable/table_apply_test.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/bigtable/table_apply_test.cc:50]: Same rationale, the function surrounding this is called many ("many", more than once..) times.

File: google/cloud/bigtable/table_check_and_mutate_row_test.cc (hunk 7 above)
Comment (coryan) [reply google/cloud/bigtable/table_check_and_mutate_row_test.cc:49]: Same rationale, though this is only two times.

**Review by coryan — COMMENTED** [review]
PTAL

**Review by dbolduc — APPROVED** [review]
My bad. I realize now that the goal is to create these fixtures as few times as possible.

**Pull request merged**